(GH-20) Add cChocoConfig resource

mkevenaar opened the pull request:
This resource allows users to change Chocolatey Config
settings using an DSC resource.

Fixes #20 

replaces PR #132

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+        $Ensure='Present',
+
+        [parameter(Mandatory = $false)]
+        [System.String]
```
Can we just use `[string]` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+    param
+    (
+        [parameter(Mandatory = $true)]
+        [System.String]
```
Can we just use `[string]` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+    param
+    (
+        [parameter(Mandatory = $true)]
+        [System.String]
```
Can we just use `[string]` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+        $ConfigName,
+
+        [ValidateSet('Present','Absent')]
+        [System.String]
```
Can we just use `[string]` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+        $Ensure='Present',
+
+        [parameter(Mandatory = $false)]
+        [System.String]
```
Can we just use `[string]` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+        if ($Ensure -eq 'Present')
+        {
+            Write-Verbose "Setting choco config $ConfigName."
+            choco config set --name $ConfigName --value $Value
```
What if `$value` is empty? We don't appear to be testing for that.

Can we also use the [Scripting Best Practices](https://chocolatey.org/docs/commands-reference#scripting-integration---best-practices-style-guide) and use `--name="'$ConfigName'" --value="'$Value'"` ?

mkevenaar replied:
I am checking the value of `$value` here: https://github.com/chocolatey/cChoco/pull/145/files#diff-6ff49aeca70224dbfb6b7427636392dfR127

pauby replied:
I may be missing it but all that is being checked there is that the parameter `Value` has been passed. There is no check that it's empty or `$null`?

mkevenaar replied:
I have updated this just now and added tests for an empty string and $null

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+        else 
+        {
+            Write-Verbose "Unsetting choco config $ConfigName."
+            choco config unset --name $ConfigName
```
Amend to use Scripting Best Practices?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+function Test-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Boolean])]
```
Can we just use `Boolean` or `Bool` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+function Get-TargetResource
+{
+    [CmdletBinding()]
+    [OutputType([System.Collections.Hashtable])]
```
Can we just use `Hashtable` here?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+
+	if($env:ChocolateyInstall -eq "" -or $null -eq $env:ChocolateyInstall)
+	{
+		$exe = (get-command choco).Source
```
Can we capitalise the `Get-Command` and use the full parameter names?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+	if($env:ChocolateyInstall -eq "" -or $null -eq $env:ChocolateyInstall)
+	{
+		$exe = (get-command choco).Source
+		$chocofolder = $exe.Substring(0,$exe.LastIndexOf("\"))
```
Is there a reason we can't use the Split-Path cmdlets here (for simplication which I always prefer)? 

Can you also do some testing to make sure you are getting the results you need? What happens for example if `Get-Command -Command choco` returns nothing or throws? Same with `$chocofolder`?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+		$chocofolder = $env:ChocolateyInstall
+	}
+	$configfolder = "$chocofolder\config"
+	$configfile = Get-ChildItem $configfolder | Where-Object {$_.Name -match "chocolatey.config$"}
```
Can we use the full parameter names for `Get-ChildItem`?

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+
+    # There is currently no choco command that only returns the settings in an CSV format.
+    # choco config list -r shows settings, sources, features and a note about API keys.
+    $xml = [xml](Get-Content $configfile.FullName)
```
Full parameter names for `Get-Content`? What do we do if the file doesn't exist or cannot be read? We're not checking for that.

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+	{
+		return $true
+	}
+}
```
Formatting is a best messed up here.

pauby commented on DSCResources/cChocoConfig/cChocoConfig.psm1:
```diff
@@ -0,0 +1,174 @@
+		}
+	}
+
+	if($Ensure -eq 'Present')
```
Can you add some comments here to clarify this (I understand it, I just think it needs some explanation for future)?

pauby reviewed (changes requested):
Just some minor changes to be made.

Can I ask for a few more comments in the code as well?

mkevenaar commented:
@pauby I have updated the code with your feedback.

pauby reviewed (commented):
This awesome. Love the comments. Just one more question left.

mkevenaar committed:
(GH-20) add cChocoConfig resource

This resource allows users to change Chocolatey Config
settings using an DSC resource.

mkevenaar committed:
(maint) Set pester to a max version of 4.10.1

pauby commented on AppVeyor/AppVeyorInstall.ps1:
```diff
@@ -29,8 +34,9 @@ Write-Host "Installed NuGet version '$($pkg.version)'"
+[version]$PesterVersion = '4.10.1'
 Install-Module -Name 'PSScriptAnalyzer' -Repository PSGallery -Force -ErrorAction Stop -MaximumVersion $ScriptAnalyzerVersion
-Install-Module -Name 'Pester' -SkipPublisherCheck -Repository PSGallery -Force -ErrorAction Stop
+Install-Module -Name 'Pester' -SkipPublisherCheck -Repository PSGallery -Force -ErrorAction Stop -MaximumVersion $PesterVersion
```
Why are we setting this here? What do later versions break?

mkevenaar replied:
The version 5.x of Pester has a lot of breaking changes. I was unable to get Pester 5 working with the current code. 

For breaking changes see: https://github.com/pester/Pester/releases/tag/5.0.0#breaking-changes

pauby commented:
Awesome work! Thanks for your contribution!
